Refactor of the fetching system

LilianBoulard opened the pull request:
Follows #147 

<hr />

This is a refactor of the dataset fetching system.
Historically, datasets were fetched from various websites.

Our aim with this update is to only use OpenML.org's API, through Scikit-learn's [`fetch_openml()`](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_openml.html) function.
This allows us to have a much more reliable and unified interface, and avoids losing access to datasets due to deletion, renaming, etc.

For instance, with the current system, 4 on the 7 datasets are unavailable (403 Access Denied, website down, etc.).

The user-interface stays the same with the functions `fetch_*()` (e.g `fetch_open_payments()`), still returning a similar dictionary.
The major difference is that this dictionary returns, among other information, a `path`, where a CSV file is located, and must be loaded (using for instance pandas' `read_csv()` function).

<hr />

TL;DR of the previous thread:
the way `fetch_openml()` is used here makes pandas a requirement.

LilianBoulard commented:
So, little info on the (I hope last) error we got here:
The test gets stopped on a fetch, which raises this error:
```
raise JSONDecodeError("Expecting value", s, err.value) from None
```
Now printing the values, we get `s=This is not JSON-encoded data!`.
This is the expected behavior: as it doesn't receive JSON-encoded data, it raises an error.
Now the thing that I don't quite understand is how does this string ends up here...

It comes from a mock at [this line](https://github.com/dirty-cat/dirty_cat/pull/155/files#diff-8f0f3ae7f57ff22fa657e96be3abe55cce37d98be950884e0d981173752436f3R167) (which is contained using a `with` statement), but the test raises at [this line](https://github.com/dirty-cat/dirty_cat/pull/155/files#diff-8f0f3ae7f57ff22fa657e96be3abe55cce37d98be950884e0d981173752436f3R59) (another function entirely), so I'm assuming this is a mock leak.

I'm investigating the issue :)

LilianBoulard commented:
So, for this (I hope last) error, we got an issue with the JSON decoding. 
It gets raised inside `sklearn.datasets.fetch_openml`, and unfortunately, this is linked to the version, as the tests pass with `scikit-learn>=0.21.0`.

Unless told to, I won't try to dig inside sklearn's code to figure out at which step it goes wrong, and perhaps monkeypatch a solution.
So meanwhile, we have two options: 
- bump sklearn's minimal version for dirty_cat to 0.21.0
- raise a clean error inviting the user to upgrade his version if he tries to fetch a dataset

What do you suggest ?

GaelVaroquaux commented:
Is this error happening for all the datasets, or only some?

GaelVaroquaux commented:
By the way, I checked, sklearn 0.21.0 is a couple years old. It's fine to bump the requirement.

LilianBoulard commented:
> Is this error happening for all the datasets, or only some?

All of them, as the error is linked to `fetch_openml` and they all use it. See this screenshot:
![image](https://user-images.githubusercontent.com/33222654/126795830-108720ee-ed23-4132-b859-156fd0299557.png)

LilianBoulard commented:
> By the way, I checked, sklearn 0.21.0 is a couple years old. It's fine to bump the requirement.

Great ! I'll make a separate PR

LilianBoulard commented:
Finally the tests pass !

The current error on CircleCI seems to be caused by insufficient memory : https://support.circleci.com/hc/en-us/articles/115014359648-Exit-code-137-Out-of-memory

For the record, I ran benchmarks on the notebooks and here are the results :

Notebook 1: 82.694251MB
Notebook 2: 149.632461MB
Notebook 3: 174.760288MB
Notebook 4: 88.800206MB
Notebook 5: 820.054484MB
Notebook 6: 1331.590108MB
Total: ~2.643 GB

As we can see, notebooks 5 and 6 stand out in terms of memory usage.
I'm not sure whether CicleCI flushes memory between each notebook.

I don't know if that's possible, but we should add these stats to the footer on the HTML render (along with `Total running time of the script: (...)`). Note: I used `tracemalloc`.

I removed the duplicate 3rd notebook (see #179), and fixed new notebook 3, the one with the SuperVectorizer, we'll see how it goes.

GaelVaroquaux commented:
Tell me when this is ready for review

LilianBoulard commented:
So the memory issue we had with the CI comes from the download process, specifically with the dataset `traffic_violation`, which is massive (see https://www.openml.org/d/42132).
I adapted the 6th notebook to use open_payments, which is way lighter.
The only downside I see is that we lost a section on the benefit of using `float32` instead of `float64`, as there are no numeric variables in the dataset. 

For the record, it was removed in 17163b7.

GaelVaroquaux commented on dirty_cat/datasets/fetching.py:
```diff
@@ -1,493 +1,380 @@
-        dataset. Warning: ``clint`` needs to be implemented and is not in the
-        requirements for now
+# Future notes:
+# - Watch out for ``fetch_openml()`` API modifications: as of january 2021, the function is marked as experimental.
```
PEP8 coding standards tell us that lines should be no more than 79 char long

GaelVaroquaux commented on dirty_cat/datasets/fetching.py:
```diff
@@ -1,493 +1,380 @@
-    # link dead.
-    # return fetch_dataset(EMPLOYEE_SALARIES_CONFIG, show_progress=False)
 
+def fetch_openml_dataset(dataset_id: int, data_directory: Path = get_data_dir()) -> dict:
```
I think that this line has a couple of PEP8 violations:
- too long
- spaces in front and after the equal sign

You should consider using a pep8 violation highlighter such as flake8

GaelVaroquaux commented on dirty_cat/datasets/fetching.py:
```diff
@@ -1,493 +1,396 @@
+            f"Could not find the dataset {dataset_id} locally. "
+            "Downloading it from OpenML; this might take a while... "
+            "If the process is interrupted, files will be invalid/incomplete. "
+            "To fix this problem, delete the CSV file if it exists. "
```
Maybe we should print also the path of the CSV file to help the user

LilianBoulard replied:
You're right ! I don't like this message altogether, it's too vague and inaccurate, I'll change it.

GaelVaroquaux commented on dirty_cat/datasets/fetching.py:
```diff
@@ -1,493 +1,396 @@
+    if not csv_path.is_file():
+        # If the CSV file does not exist, use the dataset
+        # downloaded by ``fetch_openml()`` to construct it.
+        features = _get_features(features_gz_path)
```
This seems slightly suboptimal: we are calling _get_details which reads the json from the compressed file, and latter calling _get_features, which itself also reads the compressed file.

LilianBoulard replied:
Yes, because these are two separate files :)

GaelVaroquaux commented:
> Yes, because these are two separate files :)

Right... (me go and hide)

GaelVaroquaux commented on dirty_cat/datasets/tests/test_fetching.py:
```diff
@@ -1,238 +1,352 @@
+            #
+            # One could try to manually recreate the tree structure
+            # created by ``fetch_openml()``,  and the
+            # ``.gz`` files within in order to finish the test.
```
I'm not sure that we should do this totally silently.

I think that we should:
1. Pop a warning (or something similar)
2. Skip using some of pytests' functionality for this (for instance calling pytest.skip)

GaelVaroquaux replied:
The goal is to have some track that the test is being skipped, and not silently loose track of it

GaelVaroquaux commented on dirty_cat/datasets/tests/test_fetching.py:
```diff
@@ -1,238 +1,352 @@
+
+    fetch_openml_dataset(50, test_data_dir)
+
+    mock_download       .assert_not_called()
```
Strange formatting. I don't think that this is PEP8 (but correct me if I'm wrong)

GaelVaroquaux commented on examples/02_fit_predict_plot_employee_salaries.py:
```diff
@@ -16,23 +16,24 @@
 
 ################################################################################
 # Then we load it:
 import pandas as pd
```
I know that it's ugly, but I would rather that you:
1. print the info['read_csv_kwargs']
2. pass them explicitly to read_csv
3. add a comment saying that such operation is feasible with the syntax ...**info['read_csv_args']

The reason is that I am trying to reach out to people who do not understand this syntax

GaelVaroquaux commented on examples/04_feature_interpretation_gap_encoder.py:
```diff
@@ -16,14 +16,14 @@
 # Then we load it:
 import pandas as pd
-df = employee_salaries['data']
+df = pd.read_csv(info['path'], **info['read_csv_kwargs'])
```
Same comment on the ** syntax

GaelVaroquaux commented on examples/05_scaling_non_linear_models.py:
```diff
@@ -435,7 +435,7 @@ def encode(X, y_int, one_hot_encoder, column_transformer, rbf_sampler):
 iter_csv = pd.read_csv(
     info['path'], nrows=online_train_set_size, chunksize=batchsize,
-    skiprows=1, names=columns_names, sep='\t')
+    skiprows=1, names=columns_names, **info['read_csv_kwargs'])
```
Same comment :)

GaelVaroquaux reviewed (commented):
I finished reviewing this and made all my comments. After these small things, this will be ready for merge.

LilianBoulard commented:
Good to go !

GaelVaroquaux commented on dirty_cat/datasets/tests/test_fetching.py:
```diff
@@ -65,12 +66,14 @@ def test_fetch_openml_dataset():
+            pytest.skip(
+                "One could try to manually recreate the tree structure "
+                "created by ``fetch_openml()``,  and the "
+                "``.gz`` files within in order to finish the test."
```
I think that the content of this should go in a comment, and the message in the "skip" call should be more understandable to someone who is looking only at the logs. Something like "URLError: skipping this test as the failure might be due to an internet connection problem".

LilianBoulard replied:
Ah, I see ! I'll change that

GaelVaroquaux commented on examples/01_investigating_dirty_categories.py:
```diff
@@ -16,7 +16,14 @@
+    info['path'],
+    quotechar=info['read_csv_kwargs']['quotechar'],
+    escapechar=info['read_csv_kwargs']['escapechar'],
+    na_values=info['read_csv_kwargs']['na_values'],
```
I would litterally hard code the values: "na_values='na'"

LilianBoulard replied:
Sure, why not, it should make the code easier to read

LilianBoulard commented:
These most recent changes improve the user interface.
Now, instead of reading the CSV from the disk, it is directly loaded in memory if `load_dataframe=True`.
It also returns a namedtuple instead of a dictionary, which has an object-oriented interface.

alexis-cvetkov commented:
This PR looks good to me, I think it can be merged.
Thanks Lilian !

GaelVaroquaux commented:
I was about to merge, but don't we need a an entry summarizing things in the changelog?

alexis-cvetkov commented:
Yes you're right, I always forget that.

GaelVaroquaux commented:
Yey! Merged
